lib/Command/RecreateMasterKey.php

```diff
@@ -0,0 +1,338 @@
+
+class RecreateMasterKey extends Command {
+
+	/** @var Manager  */
+	protected $encryptionManager;
+
+	/** @var IUserManager  */
+	protected $userManager;
+
+	/** @var View  */
+	protected $rootView;
+
+	/** @var KeyManager  */
+	protected $keyManager;
+
+	/** @var Util  */
+	protected $util;
+
+	/** @var  IAppManager */
+	protected $IAppManager;
+
+	/** @var  IAppConfig */
+	protected $appConfig;
+
+	/** @var IConfig  */
+	protected $config;
+
+	/** @var ISession  */
+	protected $session;
+
+	/** @var QuestionHelper  */
+	protected $questionHelper;
+
+	/** @var Setup  */
+	protected $userSetup;
+
+	/** @var IMailer  */
+	protected $mailer;
+
+	/** @var ISecureRandom  */
+	protected $secureRandom;
+
+	/** @var IL10N  */
+	protected $l;
+
+	/** @var ILogger  */
+	protected $logger;
+
+	/** @var  */
+	protected $encryptAll;
+
+	/** @var array files which couldn't be decrypted */
+	protected $failed;
+
+	/**
+	 * RecreateMasterKey constructor.
+	 *
+	 * @param IUserManager $userManager
+	 * @param View $rootView
+	 * @param KeyManager $keyManager
+	 * @param Util $util
+	 * @param IAppManager $IAppManager
+	 * @param IAppConfig $appConfig
+	 * @param IConfig $config
+	 * @param ISession $session
+	 * @param QuestionHelper $questionHelper
+	 * @param Setup $userSetup
+	 * @param IMailer $mailer
+	 * @param ISecureRandom $secureRandom
+	 * @param IL10N $l
+	 * @param ILogger $logger
+	 */
+	public function __construct(IUserManager $userManager, View $rootView, KeyManager $keyManager, Util $util,
+								IAppManager $IAppManager, IAppConfig $appConfig, IConfig $config, ISession $session,
+								QuestionHelper $questionHelper, Setup $userSetup, IMailer $mailer,
+								ISecureRandom $secureRandom, IL10N $l, ILogger $logger) {
+
+		parent::__construct();
+		$this->userManager = $userManager;
+		$this->rootView = $rootView;
+		$this->keyManager = $keyManager;
+		$this->util = $util;
+		$this->IAppManager = $IAppManager;
+		$this->appConfig = $appConfig;
+		$this->config = $config;
+		$this->session = $session;
+		$this->questionHelper = $questionHelper;
+		$this->userSetup = $userSetup;
+		$this->mailer = $mailer;
+		$this->secureRandom = $secureRandom;
+		$this->l = $l;
+		$this->logger = $logger;
+	}
+
+	protected function configure() {
+		parent::configure();
+
+		$this
+			->setName('encryption:recreate-master-key')
+			->setDescription('Replace existing master key with new one. Encrypt the file system with newly created master key')
+		;
+
+		$this->addOption(
+			'yes',
+			'y',
+			InputOption::VALUE_NONE,
+			'Answer yes to all questions'
+		);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$yes = $input->getOption('yes');
+		if ($this->util->isMasterKeyEnabled()) {
+			$question = new ConfirmationQuestion(
+				'Warning: Inorder to re-create master key, the entire ownCloud filesystem will be decrypted and then encrypted using new master key.'
```

Comment:
"In order"

Reply:
not the entire filesystem but: "all the user's home storages will be decrypted".

Will external storages be affected ? I'm not sure

Reply:
Verified with external storages:
 - [x] SFTP
 - [x] ownCloud ( without encryption when added and later running the recreate master key script )